### ISSUE TITLE
ConnectionClosed for longer running Connections

### ISSUE THREAD

**Issue by bjoernhaeuser** [pika/pika#397]
Hey,

we are running a process which publishes messages to RabbitMQ. From time to time it seems like rabbitmq closes the connection out of nowhere.

Digging into the code reveals this part:
https://github.com/pika/pika/blob/master/pika/adapters/blocking_connection.py#L235

From my not so good python knowledge this swallows the exception, right?

This is the stacktrace we get:

```
File "~/.pyenv/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 540, in basic_publish
(properties, body), False)

File "~/.pyenv/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 1121, in _send_method
self.connection.send_method(self.channel_number, method_frame, content)

File "~/.pyenv/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 249, in send_method
self._send_method(channel_number, method_frame, content)

File "~/.pyenv/lib/python2.7/site-packages/pika/connection.py", line 1496, in _send_method
self._send_frame(frame.Header(channel_number, length, content[0]))

File "~/.pyenv/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 392, in _send_frame
self.process_data_events()

File "~/.pyenv/lib/python2.7/site-packages/pika/adapters/blocking_connection.py", line 218, in process_data_events
raise exceptions.ConnectionClosed()

```

We are uing pika 0.9.13 with a BlockingConnection.

Thanks in advance and thanks for making this great open source project!
Björn

**Comment by TeddyHartanto** [comment]
I had some tests that I want to run. Everything is fine when I run it in my localhost. But, when I run the tests in a server, I had exactly the same problem and situation as this one. I checked my pika version, it's `0.9.13`. I'm using `BlockingConnection` too, by the way. 

First, I tried to call `process_data_events()` periodically. It didn't work.
Second, I tried to open a new connection and channel whenever I publish a message or open a new queue. It didn't work.

@eandersson mentioned the possibility of a socket bug. So, I checked the python version, hypothesizing that maybe there IS a bug, and it was fixed in the later version of python. In the server, the python version is `2.7.3`, and in my localhost it's `2.7.12`. To test if python version is really the problem, I installed [conda](http://conda.pydata.org/docs/) and created an environment with python version `2.7.3`. I ran the tests and it passed (I can't reproduce the issue).

After all of the above attempts, I came up with another hypothesis that maybe it's a bug in my server's `rabbitmq-server`. I compared the versions: in localhost it's the latest (`3.6.5`), in my server it's `2.8.4`. To verify that this is the actual problem, I ran the tests, but I used a remote rabbitmq with a later version. Everything worked fine. So, I upgraded the `rabbitmq-server` and, lo and behold, the problem disappears!

**TL;DR:**
The solution is to upgrade your `rabbitmq-server`!
